Improve JS_NewFloat64 API consistency

chqrlie opened the pull request:
- `JS_NewFloat64` always creates a floating point value
- add `JS_NewDouble` to try and create integer values if possible
- more fixes in quickjs.h

chqrlie committed:
Improve consistency of `JS_NewFloat64` API

- `JS_NewFloat64()` always creates a `JS_TAG_FLOAT64` value
- rename `__JS_NewFloat64` as `JS_NewFloat64__` to avoid reserved name issue
- internal `js_float64()` always creates a `JS_TAG_FLOAT64` value
- add `js_int64` for consistency
- simplify `float_is_int32` and rename as `double_is_int32`
- handle `INT32_MIN` in `double_is_int32`
- add `js_double(d)` to create a `JS_TAG_FLOAT64` or a `JS_TAG_INT` value
  if possible
- add `JS_NewDouble()` API for the same purpose
- use non testing constructor for infinities in `js_atof2`
- always store internal time value as a float64

chqrlie committed:
Merge branch 'master' into new-js-float64

chqrlie committed:
use `JS_NewNumber` and `js_number` instead of `JS_NewDouble` and `js_double`

chqrlie committed:
Merge branch 'master' into new-js-float64

chqrlie committed:
Small fixes in quickjs.h header

- remove `__` identifier prefixes
- add `JS_EXTERN` on all API functions

saghul commented on quickjs.c:
```diff
@@ -1309,11 +1285,35 @@ static JSValue js_int32(int32_t v)
 {
     if (v <= INT32_MAX)
         return js_int32(v);
-    return js_float64(v);
```
Personal preference, but I find the extra else unnecessary

chqrlie replied:
The `else` underscores the symmetry.  I could also use a ternary operator, but it is not necessarily more readable.

saghul commented on quickjs.h:
```diff
@@ -665,9 +670,9 @@ JS_EXTERN JSValue JS_GetPropertyStr(JSContext *ctx, JSValue this_obj,
-int JS_SetPropertyInternal(JSContext *ctx, JSValue this_obj,
-                           JSAtom prop, JSValue val,
-                           int flags);
+JS_EXTERN int JS_SetPropertyInternal(JSContext *ctx, JSValue this_obj,
```
Why expose this, if it's labeled internal?

chqrlie replied:
What is the meaning of this `JS_EXTERN`  tag?  My understanding is that these symbols are made visible for programs using a dynamic link version of the library.  `JS_SetPropertyInternal` is used by the `JS_SetProperty` inline function, hence must be a public symbol if we compile quickjs as a dynamic library.
Note that since the library is essentially compiled from a single source file, we avoid the classic issue regarding *visible* and *hidden* global symbols.
Whether `JS_SetProperty` should be implemented inline or just be an entry point is another question that is worth investigating.  A number of functions, macros and types currently exposed in **quickjs.h** should probably be made private to the implementation.

saghul replied:
Again, unrelated to the JS_NewNumber work, so I'd rather have it as a separate PR.

That said, I see all the other JS_SetPropertyXXX are small functions in the C file which end up calling JS_SetPropertyInternal with JS_PROP_THROW, so I think the best course here would be to drop JS_SetPropertyInternal altogether from the header file and implement JS_SetProperty in the C file.

saghul commented on quickjs.h:
```diff
@@ -562,30 +567,30 @@ JS_EXTERN JSValue JS_GetException(JSContext *ctx);
     }
 }
-JS_EXTERN void __JS_FreeValueRT(JSRuntime *rt, JSValue v);
+JS_EXTERN void JS_FreeValueRT__(JSRuntime *rt, JSValue v);
```
I don't understand the name change, if we are not moving it to the C file.

chqrlie replied:
Identifiers starting with `__` are reserved, hence the name change.

`JS_FreeValueRT` and `JS_FreeValue` are inline functions that test whether the value is refcounted.  I am not sure we should make them simple entry points and only implement them inline for the implementation.

saghul replied:
I'd leave that change for a subsequent PR, since it's unrelated to the JS_NewNumber changes.

chqrlie replied:
> I'd leave that change for a subsequent PR, since it's unrelated to the JS_NewNumber changes.

Yes, I should make separate pull requests.  This work in progress is only half baked.

chqrlie committed:
Fix CI failure

- avoid simple cycle in Array.prototype.join
- disable V8 test mjsunit/array-join.js

saghul committed:
Set version to 0.4.0

saghul committed:
Now working on 0.5.0

saghul commented on quickjs.c:
```diff
@@ -1309,11 +1285,35 @@ static JSValue js_int32(int32_t v)
+        return js_float64(v);
+}
+
+#define JS_NewInt64(ctx, val)  js_int64(val)
```
Why is this necessary, if the function is already declared in the header file?

chqrlie replied:
The external APIs should not be called from the implementation to avoid potential overhead.  I did not change all calls to `JS_NewInt64` in the code, the macro takes care of it.  I should define equivalent macros for other `JS_NewXXX` with an explanatory comment.  This is work in progress.

saghul committed:
Fix not building qjsc during release process

saghul committed:
Set version to 0.4.1

saghul committed:
Add wasi release job dependency

saghul committed:
Now working on 0.5.0

chqrlie committed:
Reverse changes unrelated to JS_NewFloat64

saghul commented:
Looks like a rebase / merge went wrong here...

chqrlie commented:
> Looks like a rebase / merge went wrong here...

Yes, some of the git commands are so confusing.  I rebased my branch on master and your 6 commits came along... How can I fix this?
I could create a new pull request from a single squashed commit, but the discussion would be lost.

saghul commented:
This is what I'd do:

Make sure your master branch is up to date.

Do an interactive rebase with it from yours, and drop any unnecessary commits: git rebase -i origin/master

Force-push the result: git push --force

bnoordhuis commented:
Instead of wholesale `git push --force`, you may want to use `git push <your-fork> +HEAD:new-js-float64` to only force-push a single branch.

chqrlie commented:
I am going to submit a new patch from a clean new fork.
